### DISCUSSION TITLE
isMenuPresented using ObservableObject...

### DISCUSSION THREAD

**Discussion by ghost** [orchetect/MenuBarExtraAccess#6] [Q&A] [answered]
### Bug Description, Steps to Reproduce, Crash Logs, Screenshots, etc.

Hello.
First of all, thank you for making a good library.
I found something while using the functions of the library, so I left a post.

I didn't want to keep passing the variable isMenuPresented to the subview with Binding for MenuBarExtraAccess, so I tried to use it wrapped in an ObservableObject.
However, using it in this way caused menubar to close and open abnormally slow (state -> binding is normal)
Below is the code of the way I implemented it.
Is Presented for controlling MenuBarExtraAccess, but is ObservableObject hard to use?

```swift
class AppState: ObservableObject {
    @Published var isMenuPresented: Bool = false
}
```

```swift
struct MyApp: App {
    @NSApplicationDelegateAdaptor(AppDelegate.self) var appDelegate
    @StateObject var appState = AppState()
    
    var body: some Scene {
        MenuBarExtra("ListeningDogApp") {
            MenuBarExtraView()
                .environmentObject(appDelegate)
                .environmentObject(appState)
        }
        .menuBarExtraStyle(.window)
        .menuBarExtraAccess(isPresented: $listengDogAppState.isMenuPresented) { statusItem in // <-- the magic ✨
            // access status item or store it in a @State var
        }
    }
}
```

```swift
struct PreferencesView: View {
    @EnvironmentObject var appDelegate: AppDelegate
    @EnvironmentObject var AppState: AppState
    @State private var isHovered = false
    
    var body: some View {
        Button {
            appState.isMenuPresented = false
            appDelegate.showMainWindow()
        } label: {
            HStack {
                Text("Preferences...")
                Spacer()
            }
            .frame(height: 30)
        }
        .buttonStyle(.plain)
    }
}
```

https://github.com/orchetect/MenuBarExtraAccess/assets/48409434/2481ad5d-c7b2-47a8-89b4-8848a5899aac

**Comment by orchetect** [comment, marked as answer]
Not sure what would explain the delay. Perhaps it's because you're using a non-English system language. The library does a lot of traversal under the hood of English property names and maybe it's having to use a redundancy to perform the action and taking longer somehow. Or perhaps something is happening between `MenuBarExtraView()` and `PreferencesView()`.

I just made a new macOS SwiftUI app with the following, and clicking the button in the menubar window makes the window dismiss immediately.

```swift
@main struct MBETestApp: App {
    @StateObject var appState: AppState = .init()
    
    var body: some Scene {
        MenuBarExtra("ListeningDogApp") {
            ContentView()
                .environmentObject(appState)
        }
        .menuBarExtraStyle(.window)
        .menuBarExtraAccess(isPresented: $appState.isMenuPresented) { statusItem in
            // access status item or store it in a @State var
        }
        
        WindowGroup {
            ContentView()
        }
    }
}

class AppState: ObservableObject {
    @Published var isMenuPresented: Bool = false
}

struct ContentView: View {
    @EnvironmentObject var appState: AppState
    
    var body: some View {
        VStack {
            Button("Toggle") {
                appState.isMenuPresented.toggle()
            }
        }
        .padding()
    }
}
```

**Comment by ghost** [comment]
I'll have to check the other part of my app again. Thank you for your answer. Have a nice day.
